dataflow,sources: fix logic that filters restored (persisted) timstamp bindings

aljoscha opened the pull request:
### Motivation

Fixes #9687

### Tips for reviewer

The first few commits just change logging a bit, we can either merge them or remove them.

The second to last commit is the real fix, and I believe this provokes the problem that the last commit then fixes.

The commits have meaningful description and there's useful comments in the code.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [N/A] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

aljoscha commented:
When you enable more logging using sth like [1] and run `TD_TEST="*failpoint*" ./mzcompose run kafka-sources` without the fix, you can see in the log that it filters out some timestamp bindings on the second (of three) startups of materialize. Thats what this fix is meant to, well, fix.

[1]
```
From 811d4e23575e62313451942a88481c8d4dfe684a Mon Sep 17 00:00:00 2001
From: Aljoscha Krettek <aljoscha.krettek@gmail.com>
Date: Wed, 19 Jan 2022 16:23:28 +0100
Subject: [PATCH] enhance logging and make timestamp frequency explicit

---
 test/persistence/mzcompose.py | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/test/persistence/mzcompose.py b/test/persistence/mzcompose.py
index 2acb24fd0..34dfbff50 100644
--- a/test/persistence/mzcompose.py
+++ b/test/persistence/mzcompose.py
@@ -19,9 +19,9 @@ from materialize.mzcompose.services import (
     Zookeeper,
 )

-mz_options = "--persistent-user-tables --persistent-kafka-upsert-source --disable-persistent-system-tables-test"
+mz_options = "--log-filter=dataflow::source=trace,persist::operators::stream=trace --persistent-user-tables --persistent-kafka-upsert-source --disable-persistent-system-tables-test"

-mz_default = Materialized(options=mz_options)
+mz_default = Materialized(timestamp_frequency="100ms", options=mz_options)

 mz_logical_compaction_window_off = Materialized(
     options=f"{mz_options} --logical-compaction-window=off"
--
2.34.1
```

ruchirK commented on src/dataflow/src/source/mod.rs:
```diff
@@ -1766,6 +1795,8 @@ impl SourceReaderPersistence {
                 .or_default() += diff;
         }
 
+        log::debug!("Lowest restored bindings ts: {}", lowest_bindings_ts);
```
couldn't this binding actually correspond to a zero diff?

aljoscha replied:
It could, I added this debug message to see what's happening with compaction, and if the lowest message we get is actually beyond the seal upper, which would be bad. I think we won't keep all the logging changes, that's why I put them out in separate commits.

ruchirK commented on src/dataflow/src/source/mod.rs:
```diff
@@ -1766,6 +1795,8 @@ impl SourceReaderPersistence {
                 .or_default() += diff;
```
I know this is not the current change, but I'm struggling to convince myself that always compacting all of the bindings up like this is correct because my understanding is that this compaction is as aggressive as possible, and will not respect the since frontier, which could be held back if there was an exactly once sink? its also possible that I'm missing something here

aljoscha replied:
It's complicated. :sweat_smile: But, as always, it's good that you ask, because it forces me to re-think, and condense my thoughts such that someone else now also knows how this works.

The core reason is that compacting a bunch of differential updates (aka advancing the since) and compacting the timestamp bindings (in the history) are different things.

I'll try and explain using an example. For different steps, we will both look at the current contents of the timestamp histories and how that is reflected in differential updates that can be used to restore that state.

(`tx` is `timestamp x`, `ox` is `offset x`, `ctx` is `capability time x`, note how the time at which updates happen is decoupled from the time that timestamp bindings refer to)

```
State 0:
ts_histories: [(t10, o5), (t20, o10)]
all emitted differential updates:
((t10, o5), ct0, 1)
((t20, o10), ct0, 1)
```

let's allow compaction up to `t30`.
This fill first advance all times, yielding `ts_histories: [(t30, o5), (t30, o10)]`, and then they will be combined to `ts_histories: [(t30, o10)]`
```
State 1:
ts_histories: [(t30, o10)]
all emitted differential updates:
((t10, o5), ct0, 1)
((t20, o10), ct0, 1)
((t10, o5), ct1, -1)
((t20, o10), ct1, -1)
((t30, o10), ct1, 1)
```

We have to full fidelity of changes to the timestamp histories in the differential updates. If we want to restore the state up to time `c1`, we would have to "sum up" all the updates that are `<=c1`, ignoring their timestamp. Same goes for restoring at `c0`. And we don't care about the compaction frontier (aka since) here, because that is implicit in the timestamp bindings that are still present.

Side note: allowing compaction (i.e. advancing the since) on the collection that stores the differential binding updates means we now loose historical resolution. If we allowed compaction up to `ct1`, the `-1`/`1` would anihilate. But it's not the same thing as "compacting" the timestamp histories.

Maybe that helps, but I'm very happy to do another call.

ruchirK commented on src/dataflow/src/source/mod.rs:
```diff
@@ -1361,23 +1375,32 @@ where
+                    // bindings here.
+                    let existing_binding = timestamp_histories
+                        .get_bindings_in_range(lower.borrow(), upper.borrow()).into_iter()
+                        .filter(|(pid, _ts, _offset)| *pid == source_ts.partition).next();
```
what if the binding in timestamp_histories actually corresponds to a lower offset, would we still want to keep it instead of the binding from persist? or is that not possible?

ruchirK replied:
ping @aljoscha

aljoscha replied:
That would be an invalid state. I don't think it can happen, though, but I will add a filter condition.

ruchirK reviewed (commented):
I think this is all starting to make more sense though I can't say that I fully understand the bug yet.

What it seemed like was happening (to me):

- we read some data and had some bindings up to some t before we triggered the failpoint
- after that, nothing was written down in persist, but some timestamp bindings were still written down in sqlite
- when we restart we somehow start reading from the wrong offset, and miss out on a subset of the data?

but I'm not really seeing where in this code we would change our start offsets. I'll keep reading and trying to understand but wanted to make sure i got my first set of comments out asap

aljoscha commented:
@ruchirK The way the bug manifests is quite subtle, I only figured out the exact workings today. This is how it works:

NOTE: It's important to keep in mind that at no point during the test do we have bindings in SQLite, because we have no sinks.

We start `materialized` three times, so we'll look at what happens for each one.

First invocation:
 - read until offset `read_offset_1`
 - successfully persist/seal bindings up to `bindings_offset_1`, which is `< read_offset_11, at seal timestamp `2`
 - successfully persist/seal data up to `data_offset_1`, which is `< bindings_offset_11, at seal timestamp `1`
 - failpoint triggers

Second invocation (this one is very short, by test design):
 - notice that the common upper seal timestamp is `1`
 - read timestamp binding `(bindings_offset_1, 2, 1)`, now the bug that `dataflow,sources: fix logic that filters restored (persisted) timstamp bindings` fixes triggers: we _think_ we have a newer binding in the histories, even though it's just the proposer, so we don't add this binding to the `timestamp_histories`, **which stay empty**
 - this will lead to the logic that differentialises the contents of the `timestamp_histories` to emit a retraction for that binding
 - we also retract some of the updates that were written but not sealed
 - we manage to successfully seal up to some seal timestamp `3`

At this point we have in the timestamp bindings, among others, these two: `(bindings_offset_1, 2, 1)` and `(bindings_offset_1, 2, -1)`, and some bindings for lower offsets at lower timestamps.

Now, on the third invocation a bug that I have fixed in https://github.com/MaterializeInc/materialize/pull/10145 triggers.

Third invocation:
 - Notice that upper seal timestamp is `3`
 - Before the fix in https://github.com/MaterializeInc/materialize/pull/10145, we don't consolidate the bindings before determining the start offset
 - We see `(bindings_offset_1, 2, -1)` (or the `+1` variant) and gather from that that the start offset should be `bindings_offset_1`, even though it is bigger than `data_offset_1`
 - we start reading from that offset, thus skipping some records that we should read

In https://github.com/MaterializeInc/materialize/pull/10145 I added a fix for the restore logic to correctly "sum up" all the bindings on restore, also before we determine starting offsets. Because of the faulty restore logic, that ignores timestamp bindings even though it shouldn't, all the bindings annihilate and we have no bindings. Meaning that with this fix, on the third invocation we read all the records again.

Third invocation (with fixed restore logic):
 - Notice that we have no bindings
 - Read all data from Kafka

And then the fix to the filtering logic (in this PR) will bring us to a correct state.

philip-stoev reviewed (approved):
I have nothing to add to this PR -- it appears that the Aljoscha was able to reproduce the problem at will, so my involvement here would be limited.

Given the success (he he) of this test I will be adding more failpoints to the code to get it to crash in yet more types of situations.

aljoscha commented:
Btw, this PR is not final, there are still instabilities (but I know where they come from).

@philip-stoev And yes, the test was super helpful in unearthing a slight problem in the design! :tada:

ruchirK commented on src/dataflow/src/source/mod.rs:
```diff
@@ -1878,6 +1878,59 @@ impl<K: Codec, V: Codec> PersistentTimestampBindingsConfig<K, V> {
+            (
+                binding,
+                bindings_cap.time().clone(),
+                diff.try_into()
```
you should be able to use ore::CastFrom

ruchirK replied:
did CastFrom not work?

aljoscha replied:
Ah no, it does work.

ruchirK commented:
I think the explanation and code changes so far make sense: I'm trying to write down invariants that need to hold

1. Partitions are never destroyed. Ie -- if you had `(bindings_offset_2, t2, 1)` and then you retracted it with `(bindings_offset_2, t2, -1)` then there must be some other `(bindings_offset_{x}, t2, +1)`. What I'm trying to say is that the collection can't go from having a nonzero number of bindings at `t1` to no bindings at `t2`. I don't think this would have prevented the issue?

> We see (bindings_offset_1, 2, -1) (or the +1 variant) and gather from that that the start offset should be bindings_offset_1, even though it is bigger than data_offset_1

2. Can we assert somewhere that the start offsets are == the correct data offsets?

3. On startup it seems like we do a lot of complicated machinery to make sure that the persisted collection of timestamp bindings and the in-memory representation of timestamp bindings are identical with each other. Could we assert (or debug assert) that they are actually identical maybe with a second snapshot of the persisted collection?

ruchirK commented:
Sorry for the delays I'll review this as soon as I get into the office today (which should be in about 3 hours)

ruchirK commented:
I'm reviewing this now!

ruchirK commented on src/dataflow/src/source/mod.rs:
```diff
@@ -1664,110 +1707,88 @@ impl SourceReaderPersistence {
+    /// timestamp updates and materializing/consolidating them into a `Vec`. This basically sums up
+    /// the `diff` of the bindings and collects those bindings whose `diff` is `1`.
+    ///
+    /// This restores two versions of the bindings: one as of an up to the common upper seal
```
Ok this was challenging but I think I get the idea.

Is it basically -- if we've sealed at up to `cap_t3` and we have bindings

```
((offset1, t1), cap_t1, 1)
((offset2, t2), cap_t2, 1)
((offset3, t3), cap_t3, 1)
```

then we want to keep all bindings < `cap_t3` and so ... we first store in `TimestampBindingUpdater`

```
((offset1, t1), 1)
((offset2, t2), 1)
((offset3, t3), 1)
```

(which is latest_bindings). Then we store in timestamp histories

```
((offset1, t1), 1)
((offset2, t2), 1)
```

then we call update with timestamp histories

which negates what already exists in timestamp_updater and then adds what is in timestamp_histories

leaving us with `((offset3, cap_t3), -1)`? I believe that's the logic unless I grossly misunderstood something. I guess to me it seems like while we are in `restore` we already know that `((offset3, t3)` can be thrown out?

Like, this isn't a full diff but a sketch of what I'm thinking about


```
diff --git a/src/dataflow/src/source/mod.rs b/src/dataflow/src/source/mod.rs
index 7e7a4b450..a9fc05de5 100644
--- a/src/dataflow/src/source/mod.rs
+++ b/src/dataflow/src/source/mod.rs
@@ -1734,7 +1734,7 @@ impl SourceReaderPersistence {
         assert!(self.config.upper_bindings_seal_ts >= self.config.upper_data_seal_ts);

         // Materialized version of all bindings updates.
-        let mut latest_bindings: HashMap<_, isize> = HashMap::new();
+        let mut invalid_bindings: HashMap<_, isize> = HashMap::new();

         // Materialized version of bindings updates that are not beyond the common seal timestamp.
         let mut valid_bindings: HashMap<_, isize> = HashMap::new();
@@ -1751,13 +1751,11 @@ impl SourceReaderPersistence {
                 *valid_bindings
                     .entry((source_timestamp.clone(), assigned_timestamp))
                     .or_default() += diff;
+            } else {
+                *invalid_bindings
+                    .entry((source_timestamp, assigned_timestamp))
+                    .or_default() += diff;
             }
-
-            // We consolidate all the diffs as we go. Below we collect all positive updates and
-            // return them.
-            *latest_bindings
-                .entry((source_timestamp, assigned_timestamp))
-                .or_default() += diff;
         }
```

It seems like we have to jump through all of the hoops of
- seed the updater with the state we want to throw away + the state we want to keep
- seed the history with the state we want to keep
- call update which does gives us back (updater - history) = -(state we want to throw away)
- persist that state.

in order to make this work with the updater / timestamp histories API. I think this could be simplified a bit if
- restore returned a set of bindings to keep and a disjoint set of bindings to throw away
- we exposed a function that was something like `emit_bindings` or something like that which `maybe_emit_bindings` also calls?
- We only keep the state we want to keep in `updater` and `timestamp_histories`

It's late and I might be missing something fundamental about why that doesn't work. if so, im really sorry!

aljoscha replied:
The analysis of what it does is exactly correct! I will change it to not rely on the differentializer to do the retracts, and instead return them from the code that restores them, because there we know exactly what to retract.

ruchirK commented on src/dataflow/src/source/mod.rs:
```diff
@@ -1340,15 +1344,45 @@ where
+                );
+
+                for (pid, offset) in starting_offsets {
+                    if crate::source::responsible_for(
```
this (and the subsequent uses of `responsible_for` feel like "restore returned all of the bindings, even ones we don't need and now we have to carefully avoid those. Instead, could we strip out the irrelevant bindings from what `restore` gives us right at the very beginning? Something like:

```
diff --git a/src/dataflow/src/source/mod.rs b/src/dataflow/src/source/mod.rs
index 7e7a4b450..bf701bd07 100644
--- a/src/dataflow/src/source/mod.rs
+++ b/src/dataflow/src/source/mod.rs
@@ -1334,7 +1334,7 @@ where
             });

             let result = timestamp_histories.and_then(|timestamp_histories| {
-                let (mut valid_bindings, latest_bindings) = source_persist.restore().map_err(|e| {
+                let (mut valid_bindings, mut latest_bindings) = source_persist.restore().map_err(|e| {
                     SourceError::new(
                         sql_name.clone(),
                         SourceErrorDetails::Persistence(format!(
@@ -1344,6 +1344,22 @@ where
                     )
                 })?;

+                // Only work with the bindings this worker is responsible for.
+                valid_bindings.retain(|(source_ts, _)| {
+                    crate::source::responsible_for(
+                        &id.source_id,
+                        worker_id,
+                        worker_count,
+                        &source_ts.partition)
+                });
+
+                latest_bindings.retain(|(source_ts, _)| {
+                    crate::source::responsible_for(
+                        &id.source_id,
+                        worker_id,
+                        worker_count,
+                        &source_ts.partition)
+                });
                 let starting_offsets = source_persist.get_starting_offsets(valid_bindings.iter());

                 tracing::debug!(
@@ -1361,13 +1377,7 @@ where
                 );

                 for (pid, offset) in starting_offsets {
-                    if crate::source::responsible_for(
-                        &id.source_id,
-                        worker_id,
-                        worker_count,
-                        &pid) {
                         timestamp_histories.add_partition(pid, Some(offset));
-                    }
                 }

                 // We need to sort by offset and then timestamp because `add_binding()` will not allow
@@ -1375,14 +1385,6 @@ where
                 valid_bindings.sort_by(|a, b| (a.0.offset.offset, a.1).cmp(&(b.0.offset.offset, b.1)));

                 for (source_ts, assigned_ts) in valid_bindings.iter() {
-                    if !crate::source::responsible_for(
-                        &id.source_id,
-                        worker_id,
-                        worker_count,
-                        &source_ts.partition) {
-                        continue;
-                    }
-
                     // The timestamp bindings are potentially pre-seeded by bindings that we
                     // restored from the coordinator, if/when coordinator based timestamp
                     // persistence is active.
@@ -1535,7 +1537,6 @@ where
                     worker_id,
                     &mut timestamp_bindings_updater,
                     timestamp_histories,
-                    &consistency_info,
                     bindings_cap,
                     bindings_output,
                 );
@@ -1629,7 +1630,6 @@ where
                 worker_id,
                 &mut timestamp_bindings_updater,
                 timestamp_histories,
-                &consistency_info,
                 bindings_cap,
                 bindings_output,
             );
@@ -1883,7 +1883,6 @@ fn maybe_emit_timestamp_bindings(
     worker_id: usize,
     timestamp_bindings_updater: &mut Option<TimestampBindingUpdater>,
     timestamp_histories: &mut TimestampBindingRc,
-    consistency_info: &ConsistencyInfo,
     bindings_cap: &Capability<u64>,
     bindings_output: &mut OutputHandle<
         u64,
@@ -1901,19 +1900,14 @@ fn maybe_emit_timestamp_bindings(
     let changes = timestamp_bindings_updater.update(timestamp_histories);

     // Emit required changes downstream.
-    let to_emit = changes
-        .into_iter()
-        .filter(|((source_ts, _assigned_ts), _diff)| {
-            consistency_info.responsible_for(&source_ts.partition)
-        })
-        .map(|(binding, diff)| {
-            (
-                binding,
-                bindings_cap.time().clone(),
-                diff.try_into()
-                    .expect("could not convert i64 diff to isize"),
-            )
-        });
+    let to_emit = changes.into_iter().map(|(binding, diff)| {
+        (
+            binding,
+            bindings_cap.time().clone(),
+            diff.try_into()
+                .expect("could not convert i64 diff to isize"),
+        )
+    });

     // We're collecting into a Vec because we want to log and emit. This is a bit
     // wasteful but we don't expect large numbers of bindings.
```

aljoscha replied:
:ok_hand:

ruchirK commented on src/dataflow/src/source/mod.rs:
```diff
@@ -1361,43 +1395,56 @@ where
+                                false,
+                            );
+                        },
+                        Some(existing_binding) => {
```
if the existing binding is not identical to the one in `persist` that will mess up the retraction/insertion logic right?

if say instead of `(a, b)` we get `(a, c)` from the sqlite bindings then we will end up retracting `(a, b)` and inserting `(a, c)`. Maybe that's fine I just wanted to call it out and see if was indeed fine?

aljoscha replied:
It is fine, yes. The retraction logic will see that the current contents of `timestamp_histories` does not match what it thinks is stored in persistence, and will retract updates to bring persist in line with the current bindings.

I think. :crossed_fingers: (it will be much easier once we don't allow mixing persist and SQLite...)

ruchirK commented on src/dataflow/src/source/mod.rs:
```diff
@@ -1474,6 +1521,26 @@ where
                 }
             };
 
+            if bindings_cap.time() == &0 {
```
this is triggering on the capability being zero but is the intention really: please run this once before anything else happens? just want to double check

aljoscha replied:
Yes, that is correct. I don't like it, but I didn't know how else to do it. I could do it if I rewrote the source code to use a custom operator, where we can just emit a thing immediately when starting up. But that seems more complicated.

ruchirK reviewed (approved):
I think I finally get everything here! Your overview on tuesday really helped. Sorry for my delay in reviewing. I left comments but they are mostly clarifying questions and potential simplifications. otherwise everything lgtm!

aljoscha commented on src/dataflow/src/source/mod.rs:
```diff
@@ -1857,6 +1878,59 @@ impl<K: Codec, V: Codec> PersistentTimestampBindingsConfig<K, V> {
     }
 }
 
+fn maybe_emit_timestamp_bindings(
```
will do!

aljoscha commented on src/dataflow/src/source/mod.rs:
```diff
@@ -1857,6 +1878,59 @@ impl<K: Codec, V: Codec> PersistentTimestampBindingsConfig<K, V> {
+    let to_emit = changes
+        .into_iter()
+        .filter(|((source_ts, _assigned_ts), _diff)| {
+            consistency_info.responsible_for(&source_ts.partition)
```
This is left over from when we didn't. I'll check again and see if we still need it. I might upgrade it to panic, in case we have something that we're not responsible for, as a sanity check.

ruchirK replied:
did you still want to upgrade this to a panic?

aljoscha commented on src/dataflow/src/source/mod.rs:
```diff
@@ -1664,110 +1707,88 @@ impl SourceReaderPersistence {
+
+    // Sums up the diffs and yields only those bindings that have a `diff` of `1`. Panics if a
+    // consolidated `diff` is not `1` or `0`.
+    fn consolidate_bindings<T: Debug>(&self, input: impl Iterator<Item = (T, isize)>) -> Vec<T> {
```
Yep, the consolidation is happening outside the call. :see_no_evil: Will change this.

aljoscha commented:
Thanks for the thorough review! I'll take next Monday to try and simplify the startup/retraction logic.

aljoscha commented:
@ruchirK I refactored/simplified this. Could you maybe take another look?

ruchirK commented on src/dataflow/src/source/mod.rs:
```diff
@@ -1346,15 +1352,47 @@ where
 
-                for (pid, offset) in offsets {
-                    timestamp_histories.add_partition(pid, offset);
+                valid_bindings.retain(|(source_ts, _assigned_ts)| {
```
maybe add a comment here explaining why we're doing this?

ruchirK commented on src/dataflow/src/source/mod.rs:
```diff
@@ -1670,109 +1717,54 @@ impl SourceReaderPersistence {
-    /// beyond the `upper_bindings_seal_ts`. This should be used when restoring to clean up unsealed
-    /// updates from previous attempts.
-    fn internal_restore(
+    /// NOTE: Consolidated bindings with a `diff` other than `0` or `1` indicate a but, and this
```
nit: typo

> Consolidated bindings with a `diff` other than `0` or `1` indicate a but,

s/a but,/an error?

ruchirK commented on src/dataflow/src/source/mod.rs:
```diff
@@ -1784,9 +1776,36 @@ impl SourceReaderPersistence {
+        // For retractions, we just need the "consolidated" diff value for each binding/timestamp.
+        let retractions = retractions
+            .drain()
+            .filter(|(_binding, diff)| diff != &0)
```
should we do the same check that the diff == 1 here?

aljoscha replied:
I'll add it, but it has to be `diff.abs()` because it can be negative. :+1:

ruchirK reviewed (approved):
LGTM! left some small nits / pings on a few prior comments / requests for further docs but after this feel free to merge whenever you want

aljoscha commented:
@ruchirK Thanks for taking another (thorough) look. I addressed the nits and will merge once ci is green. :tada:
